### PR TITLE
JP-3327 Mrs spectral leak

### LINKED ISSUE #7786: Apply MRS spectral leak correction
_Issue [JP-3327](https://jira.stsci.edu/browse/JP-3327) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

As has long been known, the MRS has a spectral leak in which 6 micron light leaks into the 12 micron channel.  A correction for this has long been part of the pipeline design and spec3 workflow (see <https://outerspace.stsci.edu/display/JWSTCC/Vanilla+Spectral+Leak+Subtraction),> but it is only now with the recent overhaul of MRS flux calibration that such a correction has become possible.

(Previously the leak was baked into the photom calibration reference files, while now it is not, so a correction relies only upon the spectrum of the science target and not the calibration source).

The algorithm for this already exists and has been tested during the flux calibration process and independently by some science teams.  In brief, a relative throughput reference file would have to be added to CRDS, multiplied by the Ch1B spectrum, and subtracted from the Ch3A spectrum.

### PR DESCRIPTION
**Pull request by jemorrison**
Ready for merging.

THIS STDATAMODELS update has been merged: stdatamodels: https://github.com/spacetelescope/stdatamodels/pull/228

Resolves [JP-3327](https://jira.stsci.edu/browse/JP-3327)

Closes #7786

This PR addresses adds the spectral leak step.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by hbushouse** [comment]
Is this still being worked or is it ready for a review?  Needs a change log entry.

**Comment by jemorrison** [comment]
@hbushouse this is still being worked. Not ready for review yet

**Comment by jemorrison** [comment]
With the recent change to the algorithm (2*wave) ch1 the test I ran before now matches well. In this test I ran the spectral leak step using extracted spectra created from IFU cubes with different Ch1 bands and different ch 3 bands, but always having at least ch1 Medium and ch3 short. Here are the results 
Screen Shot 2023-11-20 at 5 32 02 PM

File: jwst/spectral_leak/spectral_leak_step.py
Comment (drlaw1558) [inline jwst/spectral_leak/spectral_leak_step.py:64]: Should this and other print statements below be removed, or use the usual logging stream?

**Review by drlaw1558 — APPROVED** [review]
Science-wise this looks good to me, and is now working as expected.  We should probably clean up remaining print statements though, and the reference file will need to be delivered to CRDS ops before merging.

File: jwst/spectral_leak/spectral_leak_step.py
Comment (jemorrison) [reply jwst/spectral_leak/spectral_leak_step.py:64]: Whoops I forgot to remove that print statement. I will do that.

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:18]: Need whitespace before the ":" in all parameter entries

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:19]: This isn't just information, it's the "spectral leak reference file data model"

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:26]: whitespace

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:21]: whitespace

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:27]: not RSCD

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [reply jwst/spectral_leak/spectral_leak.py:18]: Wait, I see `sp_leak_ref` being used as an argument to a datamodels open call down below, so is this already a datamodel when it's passed in as an argument or is it just a file name? If it's already a model, why is it being loaded into a model again down below?

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:45]: "Note that the factor of 2 used above in "2*wave1b" is necessary, because we're dealing with second-order light.
  Without this, the interpolation would simply use the last point in the spectrum, which would be close enough
  to not be noticeable for BAND spectra, but can be quite different for Channel spectra."

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:43]: "Note that the factor of 2 used above in "2*wave1b" is necessary, because we're dealing with second-order light.
  Without this, the interpolation would simply use the last point in the spectrum, which would be close enough
  to not be noticeable for BAND spectra, but can be quite different for Channel spectra."

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:28]:
```suggestion
        input : model container
            List of input 1D spectral data models
```

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:32]:
```suggestion
        result : data model
            The corrected data model. This will be "input" if ...
```

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:50]:
```suggestion
                for i, x1d in enumerate(input_model):  # input_model is a Model Container
```

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:54]:
```suggestion
                        return input
```

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:64]: I don't think it's necessary to tell the user *how* you got the CH 1B, just say "Found CH 1B in input"

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:71]: Similar to above, I suggest "Found CH 1B in input data"

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:75]: Same comments here and below

File: jwst/spectral_leak/spectral_leak.py
Comment (jemorrison) [reply jwst/spectral_leak/spectral_leak.py:21]: I always forget the dang whitespace!

File: jwst/spectral_leak/spectral_leak_step.py
Comment (jemorrison) [reply jwst/spectral_leak/spectral_leak_step.py:64]: This comment was more for the developer/help desk investigator - if the program fails for some reason. But I can clean it up

**Comment by jemorrison** [comment]
@drlaw1558 @hbushouse 
I have put a comment on the top of this PR not to merge yet because we need the reference file in CRDS-test to be in the operational CRDS. I will go write the documentation since there is not hurry to merge this PR.

@hbushouse can you remind me about regression tests.  How do regression tests work if the reference file is in CRDS-test ?

**Comment by hbushouse** [comment]
> @drlaw1558 @hbushouse I have put a comment on the top of this PR not to merge yet because we need the reference file in CRDS-test to be in the operational CRDS. I will go write the documentation since there is not hurry to merge this PR.

Would it hurt anything to have the ref file in CRDS-Ops before the code is available? If it's a new ref file that isn't used by existing code, then it should be OK to deliver to Ops any time. If we don't include this code in our 10.1 delivery, then we have to wait all the way to 10.2 to deliver it. So why not deliver the ref file now to Ops, so that it can be used whenever the code gets merged?

> @hbushouse can you remind me about regression tests. How do regression tests work if the reference file is in CRDS-test ?

You need to specify "CRDS_PATH=/grp/crds/jwst/test" in the ENV_VARS section of "Build with Parameters" when you run a regtest.

**Comment by drlaw1558** [comment]
@hbushouse If we can get this in for 10.1 that'd be great.  The file is already in OPS, but my understanding is that it's only in a non-default context that needs to be merged back to the main context.  I'll open a CRDS ticket to discuss.

**Comment by jemorrison** [comment]
@drlaw1558 If we can get this file into CRDS ops by this Wednesday (a new CRDS build) that would be really helpful.

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:19]:
```suggestion
        name of the spectral leak reference file defined by datamodels.MirMrsPtCorrModel
```

File: jwst/spectral_leak/spectral_leak.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak.py:22]:
```suggestion
    ch1b : numpy array
        Input channel 1B spectrum
        
    ch3a : numpy array
        Input channel 3A spectrum
```

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:37]:
```suggestion
            The corrected data model. This will be "input" if the step is skipped,
            otherwise it will be a corrected 1D extracted spectrum that contains
            the MRS channel 3B range.
```

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:61]: I think this should be a warning message, not info, and the warning should also say that the step is being skipped. Then set the completion status to SKIPPED before returning `input`.  Same for the warning in line 54 (should say the step is being skipped, set the status to SKIPPED, and then return `input`).

File: jwst/spectral_leak/spectral_leak_step.py
Comment (hbushouse) [inline jwst/spectral_leak/spectral_leak_step.py:97]: Should be a warning message and need to set status to SKIPPED before returning `input`.

**Comment by drlaw1558** [comment]
Just adding a comment that the relevant reference file is now in the default CRDS-OPS context, so should not be a blocker for merging.

**Comment by hbushouse** [comment]
> Just adding a comment that the relevant reference file is now in the default CRDS-OPS context, so should not be a blocker for merging.

The only thing I'm waiting for now is updates by the author based on my latest review.

**Comment by hbushouse** [comment]
Regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1067

**Comment by jemorrison** [comment]
Working on a regression test that I can not get working. 
Once I have that I will push my updates

**Comment by hbushouse** [comment]
All non-MIRI regression tests involving calwebb_spec3 are now throwing an error, because CRDS doesn't know how to handle the prefetch for the MRSPTCORR ref file. There aren't any pmap entries for other instruments.

**Comment by jemorrison** [comment]
Interesting . I will look at see how this is done for other MIRI only reference files - like the residual fringe or mrsxartcorr.

**Comment by jemorrison** [comment]
@hbushouse I wrote a regression test and from that test I realized that the units in the output are not being written to the final spectrum. I need to research why this is occurring. Will push my latests work in a moment

File: jwst/spectral_leak/spectral_leak_step.py
Comment (jemorrison) [reply jwst/spectral_leak/spectral_leak_step.py:61]: Ok I have a question for you. 'input' will be a DataModelContainter.
So before I return and set the step to skip do I just do this: (I will try and make this fail and see what I need)
Just wondered if you knew off the top of your head

self.log.warning("Data sent to spectral_leak step is not an extracted spectrum. It is  {}." .format(type(x1d)))
input .meta.cal_step.spectral_leak = 'COMPLETE'.  # IS THIS VALID FOR THIS TYPE OF DATA
return input

**Comment by jemorrison** [comment]
@hbushouse @drlaw1558 I added some documentation since the last review. I also added a regression test that just calls the spectral_leak_step. I have been assuming that - by default - this step would be run as part of calwebb_spec3. When it is run as a standalone step the output extracted spectra do not have units and some other keyword  values. @tapastro has explained to me that there is a bug in datamodels and writing FIT files.  In the spectral leak step the extracted spectra are read using datamodels, corrected and then when the step finishes it uses the "under the hood" data model routines to write the data out. This is what is not working. 
For now how about I remove the regression test. For the time being this step can be run in calwebb_spec3 with no problem. Later when it is figured out the problem in writing the FITS files after this step completes we can add that regression test back in.  I was trying to avoid a regression test that called calwebb_spec3 because any  changes to cube build or extract1d could result in differences with a truth file that are not differences in the spectral leak correction. 
I am going to re-run the regression test suite with my added regression test now.

**Comment by drlaw1558** [comment]
My feeling is to prioritize getting this into the tag that's currently being wrapped up, but I defer to @hbushouse on the regression test and datamodel issue.

**Comment by hbushouse** [comment]
The new docs look like they're building/rendering OK. So that's good.

**Comment by jemorrison** [comment]
@hbushouse  
You can find the results of the regression test at: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1083/#showFailuresLink
Most of them seem related to a new NIRSpec flat.
But the first one  on the suffix - I do not understand that one. I saw the email about nsclean and suffix issues. I had forgotten to add spectral_leak to the suffix list. I added it and I think pulled in the changes made in the PR related to suffix. Do I need to do something different ?

**Comment by tapastro** [comment]
You need to add `spectral_leak` to the suffix list - if a step is run in standalone fashion, the suffix is of the form you added, i.e. `_spectralleakstep.fits`, but if the step is run as part of a pipeline and has save_results = True, the file suffix will match the alias, i.e. `_spectral_leak.fits`.

**Comment by jemorrison** [comment]
@tapastro Got it. I did not know it was trying to match the alias I made - but that now makes sense

**Comment by jemorrison** [comment]
@tapastro How come most of the aliases in steps do not have the 'step' in the name but in the suffix file they do have 'step' added ? Maybe it is because by default the output from this step if run in calspec3 will be an updated x1d.fits file ?

**Comment by tapastro** [comment]
> @tapastro How come most of the aliases in steps do not have the 'step' in the name but in the suffix file they do have 'step' added ? Maybe it is because by default the output from this step if run in calspec3 will be an updated x1d.fits file ?

That's for the two cases I mentioned above - I don't know where in stpipe this is encoded, but if a step is run in standalone fashion the file suffix is the alias (with any underscores removed) along with 'step'. I.e. if you were to run spectral_leak alone,  the output should have the suffix `_spectralleakstep.fits`. If instead you ran it as a part of spec3 and set spectral_leak.save_results=True, the step would dump its output as '_spectral_leak.fits'. If it's the last step to operate on the x1d models, the final spec3 output files with `_x1d.fits` suffixes should be identical to the spectral_leak saved outputs.

**Review by tapastro — CHANGES_REQUESTED** [review]
I think this table should probably get spectral_leak added: https://jwst-pipeline.readthedocs.io/en/latest/jwst/pipeline/calwebb_spec3.html#calwebb-spec3

**Comment by tapastro** [comment]
> @tapastro So I force the output of the spectral leak step to *x1d.fits files. I was trying to make it not so confusing to the user since they are used to using the x1d files and the only file that really changes is the one with CH3A.
> In the calwebb_spec3.py step I have right under process:
> self.spectral_leak.suffix = 'x1d'
> self.spectral_leak.save_results = self.save_results

This makes sense and I would recommend keeping it, I think the suffix list doesn't/can't know about the suffix overwrites, though, and wants to ensure that it covers the general cases, i.e. in the case of no overwrite.

**Review by hbushouse — COMMENTED** [review]
Just one (hopefully) final question ...

**Comment by hbushouse** [comment]
Did another (hopefully final) regtest run at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1090 and all failures are unrelated. So now my question is, why aren't there any differences for any of the MRS regtests?  Is this step getting applied or not?  Or is it just that the particular data sets we have in the MRS regtests don't get any correction for this?

File: jwst/spectral_leak/spectral_leak_step.py
Comment (jemorrison) [reply jwst/spectral_leak/spectral_leak_step.py:37]: @drlaw1558  could you confirm that we will run this step on all MRS data and no skip is needed. Since there is no way to tell what channels or bands we have until we get into the step the step decides if the correct channel/bands exist and then applies the correction to the ch3 data

**Comment by jemorrison** [comment]
I do have a regression test. I was waiting to see what the result of the "units" issue was. I can add the regression test in - I was just trying to get this step into the pipeline ASAP and add the regression test later.. But since it is taking some time  - I can add it now. Let me look at the regression test data that already exists. The regression test would need to have a test set that contained ch 1 B and ch3 A for this step to be run.

**Comment by jemorrison** [comment]
@hbushouse I added the regression test for the spectral leak. I looked at the current tests and none of them cover CH1B and CH3A. They cover multiple channels but they are limited to either IFUSHORT or IFULONG. We did not mix having an association with both. Before there was not really a reason and the test would take longer both detectors. I will run the regression test with this new  test now

File: jwst/spectral_leak/spectral_leak_step.py
Comment (drlaw1558) [reply jwst/spectral_leak/spectral_leak_step.py:37]: Confirm that we want it ON by default for all POINT source data.  Since the POINT/EXTENDED decision fork is within the code itself, the code would always need to run.  Presumably this won't cause an issue for any other spec modes?

**Comment by hbushouse** [comment]
@jemorrison It looks like your latest regtest run that includes the new test is at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1091/ is that correct?  And it only shows failures in the new test, due to the latest results from the pipeline version not having the TUNIT keywords in the output file (which is unrelated to this PR). So I guess that means everything looks OK for this PR now?

**Review by hbushouse — APPROVED** [review]
Everything looks good now.

**Pull request merged**
